File: Codes/Level 2/Greedy Techniques/Class 2/CircularTour.cpp
#include<bits/stdc++.h>
using namespace std;

// Not the most optimized, but accepted
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        for(int i=0; i<n; i++){
            long long sum = 0, count = 0;
            for(int j=i; j<2*n; j++){
                sum += gas[j%n];
                sum -= cost[j%n];
                if(sum<0){
                    i = j;
                    break;
                }
            }
            if(sum<0)   continue;
            return i;
        }
        return -1;
    }
};

File: Codes/Level 2/Greedy Techniques/Class 2/CircularTour.java


// most otimized solution
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        
        int extra = 0, start = 0, deficit = 0;
        
        for(int i=0; i<n; i++){
            extra = extra + (gas[i] - cost[i]);
            if(extra < 0){
                start = i + 1;
                deficit -= extra;
                extra = 0;
            }
        }
        if(extra >= deficit)    return start;
        return -1;
    }
}